File: stable/tom/library/strategy/mutraveler/Choice.java
/*
 *
 * Copyright (c) 2000-2006, Pierre-Etienne Moreau
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met: 
 * 	- Redistributions of source code must retain the above copyright
 * 	notice, this list of conditions and the following disclaimer.  
 * 	- Redistributions in binary form must reproduce the above copyright
 * 	notice, this list of conditions and the following disclaimer in the
 * 	documentation and/or other materials provided with the distribution.
 * 	- Neither the name of the INRIA nor the names of its
 * 	contributors may be used to endorse or promote products derived from
 * 	this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 **/
package tom.library.strategy.mutraveler;
import tom.library.strategy.mutraveler.AbstractMuStrategy;
import jjtraveler.Visitable;
import jjtraveler.reflective.VisitableVisitor;
import jjtraveler.VisitFailure;

/**
 * <code>Choice(v1,v2) = v1</code>    if v1 succeeds
 * <p>
 * <code>Choice(v1,v2) = v2</code>    if v1 fails
 * <p>
 * Visitor combinator with two visitor arguments, that tries to
 * apply the first visitor and if it fails tries the other 
 * (left-biased choice).
 * <p>
 * Note that any side-effects of v1 are not undone when it fails.
 */

public class Choice extends AbstractMuStrategy {
  public final static int FIRST = 0;
  public final static int THEN = 1;
  public Choice(VisitableVisitor first, VisitableVisitor then) {
    initSubterm(first,then);
  }
    
  public Visitable visit(Visitable visitable) throws VisitFailure {
    try {
      return visitors[FIRST].visit(visitable);
    } catch (VisitFailure f) {
      return visitors[THEN].visit(visitable);
    }
  }
}

File: tomgo/stable/library/sl/choice.go
package sl

import "errors"

// Choice tries First; if it fails with [ErrVisitFailure], tries Then on
// the original subject. Note: side effects of a failing First are not
// undone — this is consistent with the Java reference.
//
//	Choice(v1,v2) = v1 if v1 succeeds, otherwise v2.
type Choice struct{ abstractCombinator }

// NewChoice returns a Choice combinator, with the same null-tail
// shortcut as [NewSequence].
func NewChoice(first, then Strategy) Strategy {
	if then == nil {
		return first
	}
	c := &Choice{}
	c.initSubterm(first, then)
	return c
}

func (c *Choice) VisitLight(subject any, intro Introspector) (any, error) {
	v, err := c.args[0].VisitLight(subject, intro)
	if err == nil {
		return v, nil
	}
	if errors.Is(err, ErrVisitFailure) {
		return c.args[1].VisitLight(subject, intro)
	}
	return nil, err
}
